File: Data_Structure/Python/LinkedList.py
class Node:
    def __init__(self,data=None,next=None):
        self.data=data
        self.next=next

class LinkedList:
    def __init__(self):
        self.head=None
    def insertAtBegining(self,data):
        node = Node(data,self.head)
        self.head=node
    def insertAtEnd(self,data):
        if self.head is None:
            self.head =Node(data,None)
            return
        itr =self.head
        while itr.next:
            itr=itr.next
        itr.next=Node(data,None)
    def print(self):
        if self.head is None:
            print("The linked list is empty")
            return
        
        itr=self.head
        linkedlist_str=""
        while itr:
            linkedlist_str+=str(itr.data)+"-->"
            itr=itr.next
        print(linkedlist_str)
    def inserListValue(self,data_list):
        for  data in data_list:
            self.insertAtEnd(data)
    def lenOfLinkedList(self):
        counter =0
        itr =self.head
        if itr is None:
            print("The list is empty")
            return
        while itr:
            counter+=1
            itr=itr.next
        return counter
    def removeAtIndex(self,index):
        if index<0 or index>self.lenOfLinkedList():
            raise Exception("Invalid Index")
        if index ==0:
            self.head=self.head.next
            return
        counter=0
        itr=self.head
        while itr:
            if counter == index-1:
                itr.next=itr.next.next
                break
            itr=itr.next
            counter+=1
    def insertAtIndex(self, index, data):
        if index < 0 or index > self.lenOfLinkedList():
            raise Exception("Invalid Index")

        if index == 0: 
            self.insertAtBegining(data)
            return

        count = 0
        itr = self.head
        while itr:
            if count == index - 1:  
                node = Node(data, itr.next)  
                itr.next = node 
                break       
            itr = itr.next
            count += 1
    def insertAfterValue(self,value,data):
        itr = self.head
        while itr:
            if itr.data is value:
                node = Node(data,itr.next)
                itr.next=node
                break
            itr=itr.next
    def removeByValue(self,value):
        itr=self.head
        while itr:
            if itr.next.data is value:
                itr.next=itr.next.next
                break

ll=LinkedList()
list=["Apple","Banan","Mango","Orenge"]
ll.inserListValue(list)
ll.print()
ll.removeAtIndex(2)
ll.print()
ll.insertAtIndex(1,"Givi")
ll.print()
ll.insertAtIndex(3,"jackfruit")
ll.print()
ll.insertAfterValue("Apple",10)
ll.print()
ll.removeByValue(10)
ll.print()
print("Length of linked list:",ll.lenOfLinkedList())
        

File: Data_Structure/Java/LinkedList.java
class Node {
    String data;
    Node next;

    public Node(String data) {
        this.data = data;
        this.next = null;
    }
}

class Linkedlist {

    private Node head;
    Linkedlist() {
        this.head = null;
    }

    public void insertAtBeginning(String data) {
        Node newNode = new Node(data);
        newNode.next = head;
        head = newNode;
    }
    public void insertAtEnd(String data) {
        Node newNode = new Node(data);
        if (head == null) {
            head = newNode;
            return;
        }
        Node current = head;
        while (current.next != null) {
            current = current.next;
        }
        current.next = newNode;
    }

    public void display() {
        Node current = head;
        while (current != null) {
            System.out.print(current.data + " -> ");
            current = current.next;
        }
        System.out.println("null");
    }

    public void insertMultipleAtBeginning(String[] data) {
        for (String value : data) {
            insertAtBeginning(value);
        }
    }
    public void insertMultipleAtEnd(String[] data) {
        for (String value : data) {
            insertAtEnd(value);
        }
    }

    public int length() {
        int count = 0;
        Node current = head;
        while (current != null) {
            count++;
            current = current.next;
        }
        return count;
    }

    public void insertAtIndex(int index, String data) {
        if (index < 0 || index > length()) {
            System.out.println("Index out of bounds");
            return;
        }
        if (index == 0) {
            insertAtBeginning(data);
            return;
        }
        Node newNode = new Node(data);
        Node current = head;
        for (int i = 0; i < index - 1; i++) {
            current = current.next;
        }
        newNode.next = current.next;
        current.next = newNode;
    }

    public void removeAtIndex(int index) {
        if (index < 0 || index >= length()) {
            System.out.println("Index out of bounds");
            return;
        }
        if (index == 0) {
            head = head.next;
            return;
        }
        Node current = head;
        for (int i = 0; i < index - 1; i++) {
            current = current.next;
        }
        current.next = current.next.next;
    }

    public void insertAfterValue(String target, String data) {
        Node current = head;
        while (current != null && !current.data.equals(target)) {
            current = current.next;
        }
        if (current == null) {
            System.out.println("Value not found");
            return;
        }
        Node newNode = new Node(data);
        newNode.next = current.next;
        current.next = newNode;
    }
    public void removeByValue(String value) {
        if (head == null) {
            System.out.println("List is empty");
            return;
        }
        if (head.data.equals(value)) {
            head = head.next;
            return;
        }
        Node current = head;
        while (current.next != null && !current.next.data.equals(value)) {
            current = current.next;
        }
        if (current.next == null) {
            System.out.println("Value not found");
            return;
        }
        current.next = current.next.next;
    }
    public void clear() {
        head = null;
    }
}

public class LinkedList {
    public static void main(String[] args) {
        Linkedlist list = new Linkedlist();
        list.insertAtBeginning("Apple");
        list.insertAtEnd("Banana");
        list.insertAtEnd("Orenge");
        list.display(); 

    
        String [] fruits = {"Pineapple", "Mango", "Grapes"};
        list.insertMultipleAtBeginning(fruits);
        list.display();

        // list.insertMultipleAtBeginning(new String[]{"Jackfruite", "Mango"});
        // list.display(); 

        list.insertMultipleAtEnd(new String[]{"Givi", "Pineapple"});
        list.display(); 

        System.out.println("Length: " + list.length()); 

        list.insertAtIndex(3, "Grapes");
        list.display(); 

        list.removeAtIndex(2);
        list.display(); 

        list.insertAfterValue("Pineapple", "Papaya");
        list.display(); 

        list.removeByValue("Grapes");
        list.display(); 
    }

    public void add(int i) {
        // TODO Auto-generated method stub
        throw new UnsupportedOperationException("Unimplemented method 'add'");
    }
    
}
